agent_app_sample_code/02_data_pipeline.py

```diff
@@ -0,0 +1,421 @@
+)
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC # Data Pipeline 
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## Start MLflow run for tracking
+
+# COMMAND ----------
+
+mlflow.end_run()
+mlflow.start_run(run_name=POC_DATA_PIPELINE_RUN_NAME)
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## Load documents into a delta table
+# MAGIC
+# MAGIC In this step, we'll load files from `SOURCE_PATH` (defined in your `00_global_config`) into a delta table. The contents of each file will become a separate row in our delta table.
+# MAGIC
+# MAGIC The path to the source document will be used as the `doc_uri` which is displayed to your end users in the Agent's web application.
+# MAGIC
+# MAGIC After you test your POC with stakeholders, you can return here to change the parsing logic or extraction additional metadata about the documents to help improve the quality of your retriever.
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ### Define your parsing function
+# MAGIC
+# MAGIC This default implementation parses PDF, HTML, and DOCX files using open source libraries.  Adjust `file_parser(...)` to add change the parsing logic, add support for more file types, or extract additional metadata about each document.
+
+# COMMAND ----------
+
+from typing import TypedDict
+from datetime import datetime
+import warnings
+import io
+import traceback
+
+# PDF libraries
+from pypdf import PdfReader
+
+# HTML libraries
+from markdownify import markdownify as md
+import markdownify
+import re
+
+## DOCX libraries
+import pypandoc
+import tempfile
+
+# Schema of the dict returned by `file_parser(...)`
+class ParserReturnValue(TypedDict):
+    # DO NOT CHANGE THESE NAMES - these are required by Evaluation & Framework
+    # Parsed content of the document
+    doc_content: str  # do not change this name
+    # The status of whether the parser succeeds or fails, used to exclude failed files downstream
+    parser_status: str  # do not change this name
+    # Unique ID of the document
+    doc_uri: str  # do not change this name
+
+    # OK TO CHANGE THESE NAMES
+    # Optionally, you can add additional metadata fields here
+    example_metadata: str
+    last_modified: datetime
+
+
+# Parser function.  Replace this function to provide custom parsing logic.
+def file_parser(
+    raw_doc_contents_bytes: bytes,
+    doc_path: str,
+    modification_time: datetime,
+    doc_bytes_length: int,
+) -> ParserReturnValue:
+    """
+    Parses the content of a PDF document into a string.
+
+    This function takes the raw bytes of a PDF document and its path, attempts to parse the document using PyPDF,
+    and returns the parsed content and the status of the parsing operation.
+
+    Parameters:
+    - raw_doc_contents_bytes (bytes): The raw bytes of the document to be parsed (set by Spark when loading the file)
+    - doc_path (str): The path of the document, used to verify the file extension (set by Spark when loading the file)
```

Comment:
Will this thing always be a `dbfs:/` path / should we document that? Looks like we assume it's gonna be dbfs:/ below